src/CoreWCF.NetTcp/src/CoreWCF/Configuration/ServiceModelWebHostBuilderExtensions.cs

```diff
@@ -69,38 +60,41 @@ internal class NetTcpFramingOptionsSetup : IConfigureOptions<KestrelServerOption
     {
         private readonly ILogger<NetTcpFramingOptions> _logger;
         private readonly NetTcpFramingOptions _options;
+        private readonly IServiceBuilder _serviceBuilder;
 
-        public NetTcpFramingOptionsSetup(IOptions<NetTcpFramingOptions> options, ILogger<NetTcpFramingOptions> logger)
+        public NetTcpFramingOptionsSetup(IOptions<NetTcpFramingOptions> options, IServiceBuilder serviceBuilder, ILogger<NetTcpFramingOptions> logger)
         {
+            _options = options.Value ?? new NetTcpFramingOptions();
+            _serviceBuilder = serviceBuilder;
             _logger = logger;
-            _options = options.Value;
         }
 
+        public List<ListenOptions> ListenOptions { get; } = new List<ListenOptions>();
+
         public void Configure(KestrelServerOptions options)
         {
-            IServiceBuilder serviceBuilder = options.ApplicationServices.GetRequiredService<IServiceBuilder>();
             foreach (var endpoint in _options.EndPoints)
             {
                 options.Listen(endpoint, builder =>
                 {
                     builder.UseConnectionHandler<NetMessageFramingConnectionHandler>();
+                    var handler = builder.ApplicationServices.GetRequiredService<NetMessageFramingConnectionHandler>();
+                    // Save the ListenOptions to be able to get final port number for adding BaseAddresses later
+                    ListenOptions.Add(builder);
                 });
+            }
 
-                var baseAddress = new Uri($"net.tcp://localhost:{endpoint.Port}/");
+            _serviceBuilder.Opening += OnServiceBuilderOpening;
+        }
 
-                if (endpoint.Port == 0)
-                {
-                    // Implicit port: it hasn't been assigned by the TCP stack yet
-                    var serverFeatures = options.ApplicationServices.GetRequiredService<IFeatureCollection>();
-                    var serverAddressesFeature = serverFeatures.Get<IServerAddressesFeature>();
-                    serverAddressesFeature.Addresses.Add(baseAddress.ToString());
-                }
-                else
-                {
-                    // Explicit port: can safely add directly to the serviceBuilder (default behavior)
-                    _logger.LogDebug($"Adding base address {baseAddress} to serviceBuilder");
-                    serviceBuilder.BaseAddresses.Add(baseAddress);
-                }
+        private void OnServiceBuilderOpening(object sender, EventArgs e)
+        {
+            foreach (var listenOptions in ListenOptions)
+            {
+                var endpoint = listenOptions.IPEndPoint;
+                var baseAddress = new Uri($"net.tcp://localhost:{endpoint.Port}/");
```

Comment:
Using IServerAddressesFeature will get you every address Kestrel binds to, not just the ones added via NetTcpFramingOptionsSetup.Configure(). IServerAddressesFeature is also supported by non-Kestrel servers. I'm not sure if either of those things matter.

Reply:
For HTTP, we don't care about the which ports are used. The behavior of WCF is that when you configure using the HTTP transport, we really don't care about which port it arrived on or for what hostname. The endpoint selection code converts http and https to just http for the purposes of routing the request so if an Http binding is used, To maintain the same behavior in CoreWCF, I just add "http://localhost/" to the base addresses and don't care about what's in IServerAddressesFeature.  
This means I only care about the NetTcp listening port I've added which means I need to have a list of the ListenOptions instances which were only used for NetTcp which is why I stash. Then in the initialization of CoreWCF, I use the list of ListenOptions to add the relevant addresses to IServerAddressesFeature.  
IServerAddressesFeature is useless for my use case as it claims everything is an http endpoint when that's not the case.